sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala

```diff
@@ -0,0 +1,142 @@
+      } else if (!schemasCompatible(storedKeySchema, keySchema) ||
+        !schemasCompatible(storedValueSchema, valueSchema)) {
+        logError(errorMsg)
+        throw StateSchemaNotCompatible(errorMsg)
+      } else {
+        logInfo("Detected schema change which is compatible: will overwrite schema file to new.")
+        // It tries best-effort to overwrite current schema file.
+        // the schema validation doesn't break even it fails, though it might miss on detecting
+        // change which is not a big deal.
+        createSchemaFile(keySchema, valueSchema)
+      }
+    } else {
+      // schema doesn't exist, create one now
+      logDebug(s"Schema file for provider $providerId doesn't exist. Creating one.")
+      createSchemaFile(keySchema, valueSchema)
+    }
+  }
+
+  private def schemasCompatible(storedSchema: StructType, schema: StructType): Boolean =
+    equalsIgnoreCompatibleNullability(storedSchema, schema)
+
+  private def equalsIgnoreCompatibleNullability(from: DataType, to: DataType): Boolean = {
+    // This implementations should be same with DataType.equalsIgnoreCompatibleNullability, except
+    // this shouldn't check the name equality.
+    (from, to) match {
+      case (ArrayType(fromElement, fn), ArrayType(toElement, tn)) =>
+        (tn || !fn) && equalsIgnoreCompatibleNullability(fromElement, toElement)
+
+      case (MapType(fromKey, fromValue, fn), MapType(toKey, toValue, tn)) =>
+        (tn || !fn) &&
+          equalsIgnoreCompatibleNullability(fromKey, toKey) &&
+          equalsIgnoreCompatibleNullability(fromValue, toValue)
+
+      case (StructType(fromFields), StructType(toFields)) =>
+        fromFields.length == toFields.length &&
+          fromFields.zip(toFields).forall { case (fromField, toField) =>
+              (toField.nullable || !fromField.nullable) &&
+              equalsIgnoreCompatibleNullability(fromField.dataType, toField.dataType)
+          }
+
+      case (fromDataType, toDataType) => fromDataType == toDataType
+    }
+  }
+
+  private def readSchemaFile(): (StructType, StructType) = {
+    val inStream = fm.open(schemaFileLocation)
+    try {
+      val version = inStream.readInt()
+      // Currently we only support version 1, which we can simplify the version validation and
+      // the parse logic.
+      require(version == StateSchemaCompatibilityChecker.VERSION,
+        s"version $version is not supported.")
+
+      val keySchemaStr = inStream.readUTF()
+      val valueSchemaStr = inStream.readUTF()
+
+      (StructType.fromString(keySchemaStr), StructType.fromString(valueSchemaStr))
+    } catch {
+      case e: Throwable =>
+        logError(s"Fail to read schema file from $schemaFileLocation", e)
+        throw e
+    } finally {
+      inStream.close()
+    }
+  }
+
+  private def createSchemaFile(keySchema: StructType, valueSchema: StructType): Unit = {
+    val outStream = fm.createAtomic(schemaFileLocation, overwriteIfPossible = true)
```

Comment:
This won't overwrite an existing file if the file system doesn't support atomic overwrite. We may need to redesign this. Maybe we should write a file in each batch to avoid file overwriting.

Reply:
OK this is major one to consider. At the time I implemented this I don't know about the mess of filesystems and only see bright side. Now I see that's not going to work with some filesystems.

Writing file per batch would produce small file issues and we'll have to clean up. This is checked only once for the start of the query, so it sounds to bring pretty much overhead.

Probably it's just OK to store "initial" schema of the state (that said we don't need to overwrite), as compatibility check must be transitive. Initial schema defines the schema on state, and Spark can only write to the state if the schema of the row being written to the state is compatible. Reading the state with such schema should work as well.